File: python/day_5/part_2.py
def solve(input: str) -> int:
    sections = input.strip().split("\n\n")

    ranges = []
    for line in sections[0].strip().split("\n"):
        if line.strip():
            start, end = map(int, line.split("-"))
            ranges.append((start, end))

    ranges.sort()

    merged = []
    for start, end in ranges:
        if merged and start <= merged[-1][1] + 1:
            merged[-1] = (merged[-1][0], max(merged[-1][1], end))
        else:
            merged.append((start, end))

    total = 0
    for start, end in merged:
        count = end - start + 1
        total += count

    return total


if __name__ == "__main__":
    with open("inputs/day_5.txt", "r") as f:
        input = f.read()
        result = solve(input)
        print(result)

File: rust/day_5/part_2.rs
use std::fs;

fn solve(input: &str) -> i64 {
    let sections: Vec<&str> = input.trim().split("\n\n").collect();

    let mut ranges = Vec::new();
    for line in sections[0].trim().lines() {
        let line = line.trim();
        if !line.is_empty() {
            let parts: Vec<&str> = line.split('-').collect();
            let start: i64 = parts[0].parse().unwrap();
            let end: i64 = parts[1].parse().unwrap();
            ranges.push((start, end));
        }
    }

    ranges.sort();

    let mut merged: Vec<(i64, i64)> = Vec::new();
    for (start, end) in ranges {
        if !merged.is_empty() && start <= merged.last().unwrap().1 + 1 {
            let last_idx = merged.len() - 1;
            merged[last_idx].1 = merged[last_idx].1.max(end);
        } else {
            merged.push((start, end));
        }
    }

    let mut total = 0;
    for (start, end) in merged {
        let count = end - start + 1;
        total += count;
    }

    total
}

fn main() {
    let input = fs::read_to_string("../inputs/day_5.txt").expect("Failed to read input file");
    let result = solve(&input);
    println!("{}", result);
}
